### PR TITLE
AD incompressible navier stokes

### PR DESCRIPTION
**Pull request by lindsayad**
Last piece needed in order to move Baldr to using upstream MOOSE

### CODE REVIEW

File: framework/src/kernels/ADKernelSUPG.C [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+void
+ADKernelSUPGTempl<T, compute_stage>::computeResidual()
+{
+  prepareVectorTag(_assembly, _var.number());
+
+  precalculateResidual();
+  const unsigned int n_test = _grad_test.size();
+  for (_qp = 0; _qp < _qrule->n_points(); _qp++)
+  {
+    const auto value = precomputeQpStrongResidual() * _ad_JxW[_qp] * _ad_coord[_qp] * _tau[_qp];
```
Comment (dschwen) [inline framework/src/kernels/ADKernelSUPG.C]: shouldn't the tau be on the other term?

File: framework/include/utils/MathUtils.h [hunk 2]
```diff
@@ -77,6 +77,35 @@ addScaled(const T & scalar, const NumericVector<T2> & numeric_vector, NumericVec
...
+dotProduct(const W<T> & a, const W2<T2> & b)
+{
+  return a * b;
+}
+
+template <typename T,
+          typename T2,
+          template <typename> class W,
+          template <typename> class W2,
+          typename std::enable_if<std::is_same<typename W<T>::index_type,
```
Comment (jwpeterson) [inline framework/include/utils/MathUtils.h]: Yo dawg, heard you like nested template parameters!

**Review by jwpeterson — DISMISSED** [review]
I won't pretend to understand all the macros and template magic here, but this is definitely an awesome capability!

File: framework/src/kernels/ADKernelSUPG.C (hunk 1 above)
Comment (lindsayad) [reply framework/src/kernels/ADKernelSUPG.C]: You mean inside the `_i` loop? It's only a function of `_qp`

File: framework/include/utils/MathUtils.h (hunk 2 above)
Comment (lindsayad) [reply framework/include/utils/MathUtils.h]: 🤣

File: framework/src/kernels/ADKernelSUPG.C (hunk 1 above)
Comment (dschwen) [reply framework/src/kernels/ADKernelSUPG.C]: Oh, I see. I thought SUPG is

(test + tau * grad_test * v * adv) * strong_residual

But I'm not an SUPG expert by any stretch of the definition of the word "expert" :-D

**Review by dschwen — DISMISSED** [review]

File: framework/src/kernels/ADKernelSUPG.C (hunk 1 above)
Comment (lindsayad) [reply framework/src/kernels/ADKernelSUPG.C]: > (test + tau * grad_test * v * adv) * strong_residual

This is almost correct. For a general transient advection-diffusion problem, 

u_dot + v * grad_u + grad * (-kappa * grad_u) - f = 0

the overall weak form would look like:

(test, u_dot) + (test, v * grad_u) + (grad_test, kappa * grad_u) - (test, f) - <test, n * kappa * grad_u> + (tau * v * grad_test, u_dot + v * grad_u + grad * (-kappa * grad_u) - f) = 0

So the difference is in the treatment of second order terms...those terms are integrated by parts when applying the standard test function, but not when applying the Petrov-Galerkin modificiation, tau * v * grad_test.

So in this design, an `ADKernelSUPG` derived object is only meant to be used to add the contribution from the Petrov-Galerkin modification. Other kernels should supply the standard test function contributions.

**Review by aeslaughter — APPROVED** [review]

**Pull request merged**